rllib/algorithms/algorithm.py

```diff
@@ -2560,6 +2758,134 @@ def __setstate__(self, state: dict):
+             method and - if applicable - the "v0" checkpoint file.
+        """
+        # Do we have an old ("v0") single checkpoint file
+        # (`checkpoint_00001/checkpoint-00001`)?
+        v0_checkpoint_file = None
+
+        # `checkpoint` is a str: Could be checkpoint file (older checkpoint versions)
+        # or a directory.
+        if isinstance(checkpoint, str):
+            # Checkpoint is dir: Figure out whether this is an older checkpoint format
+            # (with a `checkpoint-\d+` file in it).
+            if os.path.isdir(checkpoint):
+                for file in os.listdir(checkpoint):
+                    path_file = os.path.join(checkpoint, file)
+                    if os.path.isfile(path_file) and re.match("checkpoint-\\d+", file):
+                        v0_checkpoint_file = path_file
+                        break
+                checkpoint = Checkpoint.from_directory(checkpoint)
+            # Checkpoint is a file: Use as-is (interpreting it as old checkpoint
+            # version).
+            elif os.path.isfile(checkpoint):
+                v0_checkpoint_file = checkpoint
+                checkpoint = Checkpoint.from_directory(os.path.dirname(checkpoint))
+            else:
+                raise ValueError(
+                    f"Given checkpoint ({checkpoint}) not found! Must be a "
+                    "checkpoint directory (or a file for older checkpoint versions)."
+                )
+
+        # Open main state file of checkpoint (or v0_checkpoint_file itself).
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            checkpoint.to_directory(tmp_dir)
+            state_file = v0_checkpoint_file or os.path.join(tmp_dir, "state.pkl")
+
+            if not os.path.isfile(state_file):
+                raise ValueError(
+                    "Given checkpoint does not seem to be valid! No file "
+                    "with the name `state.pkl` (or `checkpoint-[0-9]+`) found. "
+                    "It's possible that you have passed an older checkpoint version. "
+                    "In this case, do the following:\n"
+                    "1) Create a new Algorithm object using the original config.\n"
+                    "2) Call the `restore()` method of this algo object passing it"
+                    " your older checkpoint dir."
+                )
+
+            state = pickle.load(open(state_file, "rb"))
+
+            # New checkpoint format: Policies are in separate sub-dirs.
+            # Note: Algorithms like ES/ARS don't have a WorkerSet, so we just return
+            # the plain state here.
+            if not v0_checkpoint_file and state.get("worker") is not None:
+                worker_state = state["worker"]
+
+                # TODO: Remove filters once they are part of policies (via connectors).
+                policies = set(
+                    policies if policies is not None else worker_state["filters"].keys()
```

Comment:
this variable should really be policy_ids? name it policies is a bit confusing.
why are we relying on filters (something that will be removed) to get the list of policy keys here?
shouldn't we find them from the config dict if not specified?

Reply:
What if user added new policies later? Those new ones would NOT be part of their config, but will still be properly saved in the checkpoint.

Reply:
But yes, `policy_ids` is a better name, will change.

Reply:
oh, totally. my question was more about why do we get these IDs from the filters which we are deprecating soon-ish?
like, why not just read them out of the policy_map?

Reply:
Chicken and egg problem:
* The policy map is not known at this point, b/c you haven't created any policies yet. We are in the process - at this point in the code - to just read all the policy states that are needed (as per users `policy_ids` arg) into memory from the pickled `policy_state.pkl` files.
* So we either get the policy IDs from the names of the sub-directories OR we add a new field to the worker's state: `policy_ids`.

Reply:
I added a `policy_ids` field to the worker state and renamed the worker's "state" key into "policy_state" for clarity (still backward compatible with v0 (now: "v0.1") checkpoints).